src/plugin/shadow-plugin-tgen/shd-tgen-driver.c

```diff
@@ -312,6 +312,44 @@ static void _tgendriver_handleChoose(TGenDriver* driver, TGenAction* action) {
     _tgendriver_processAction(driver, nextAction);
 }
 
+static void _tgendriver_chooseWeightsNextAction(TGenDriver* driver, TGenAction* action){
+    TGEN_ASSERT(driver);
+
+    /* Pick a random value among all of the weights */
+    gdouble randomWeight = g_random_double_range(0, tgenaction_getTotalWeight(action));
+
+    GQueue* nextActions = tgengraph_getNextActions(driver->actionGraph, action);
+    g_assert(nextActions);
+
+    TGenAction* nextAction;
+
+    double totalWeight = 0.0;
+    /* Keep adding values from weights until we've met randomWeight */
+    while(totalWeight < randomWeight) {
+        nextAction = g_queue_pop_head(nextActions);
+        gdouble* thisWeight = tgengraph_getEdgeWeight(driver->actionGraph, action, nextAction);
+        totalWeight += *thisWeight;
+    }
+
+    _tgendriver_processAction(driver, nextAction);
```

Comment:
I think we should only call this if nextAction is not NULL, which hopefully would never happen as long as we initialize it to the first element of the queue. Still good to check anyway.